cedar-policy-core/src/entities/json/value.rs

```diff
@@ -283,53 +283,113 @@ impl CedarValueJson {
+        match value {
+            Value::Lit(lit) => Ok(Self::from_lit(lit)),
+            Value::Set(set) => Ok(Self::Set(
+                set.iter()
+                    .cloned()
+                    .map(Self::from_value)
+                    .collect::<Result<_, _>>()?,
+            )),
+            Value::Record(map) => {
+                // if `map` contains a key which collides with one of our JSON
+                // escapes, then we have a problem because it would be interpreted
+                // as an escape when being read back in.
+                check_for_reserved_keys(map.keys())?;
+                Ok(Self::Record(
+                    map.iter()
+                        .map(|(k, v)| Ok((k.clone(), Self::from_value(v.clone())?)))
+                        .collect::<Result<JsonRecord, JsonSerializationError>>()?,
+                ))
+            }
+            Value::ExtensionValue(ev) => {
+                let ext_fn: &Name = &ev.constructor;
+                Ok(Self::ExtnEscape {
+                    __extn: FnAndArg {
+                        ext_fn: ext_fn.to_string().into(),
+                        arg: match ev.args.as_slice() {
+                            [ref expr] => Box::new(Self::from_expr(expr.as_borrowed())?),
+                            [] => {
+                                return Err(JsonSerializationError::ExtnCall0Arguments {
+                                    func: ext_fn.clone(),
+                                })
+                            }
+                            _ => {
+                                return Err(JsonSerializationError::ExtnCall2OrMoreArguments {
+                                    func: ext_fn.clone(),
+                                })
+                            }
+                        },
+                    },
+                })
+            }
+        }
+    }
+
     /// Convert a Cedar literal into a `CedarValueJson`.
     pub fn from_lit(lit: Literal) -> Self {
         match lit {
             Literal::Bool(b) => Self::Bool(b),
             Literal::Long(i) => Self::Long(i),
             Literal::String(s) => Self::String(s),
             Literal::EntityUID(euid) => Self::EntityEscape {
-                __entity: (*euid).clone().into(),
+                __entity: unwrap_or_clone(euid).into(),
             },
         }
     }
 }
 
+/// helper function to check if the given keys contain any reserved keys,
+/// throwing an appropriate `JsonSerializationError` if so
+fn check_for_reserved_keys<'a>(
+    mut keys: impl Iterator<Item = &'a SmolStr>,
+) -> Result<(), JsonSerializationError> {
+    // We could be a little more permissive here, but to be
+    // conservative, we throw an error for any record that contains
+    // any key with a reserved name, not just single-key records
+    // with the reserved names.
+    let reserved_keys: HashSet<&str> = HashSet::from_iter(["__entity", "__extn", "__expr"]);
```

Comment:
Can we move this constant somewhere else instead of having it buried in a function. If we add a new reserved key we will need to update this, and that should be obvious.

Reply:
Fair enough